idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtPsiTypeProvider.kt

```diff
@@ -5,28 +5,28 @@
 
 package org.jetbrains.kotlin.idea.frontend.api.components
 
+import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiType
+import org.jetbrains.kotlin.idea.frontend.api.types.KtType
 import org.jetbrains.kotlin.load.kotlin.TypeMappingMode
-import org.jetbrains.kotlin.psi.KtDoubleColonExpression
-import org.jetbrains.kotlin.psi.KtExpression
-import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.*
 
 public abstract class KtPsiTypeProvider : KtAnalysisSessionComponent() {
-    public abstract fun getPsiTypeForKtExpression(expression: KtExpression, mode: TypeMappingMode): PsiType
-    public abstract fun getPsiTypeForKtTypeReference(ktTypeReference: KtTypeReference, mode: TypeMappingMode): PsiType
-    public abstract fun getPsiTypeForReceiverOfDoubleColonExpression(
-        ktDoubleColonExpression: KtDoubleColonExpression,
-        mode: TypeMappingMode
-    ): PsiType?
+    public abstract fun commonSuperType(types: Collection<KtType>, context: KtExpression, mode: TypeMappingMode): PsiType?
+    public abstract fun asPsiType(type: KtType, context: PsiElement, mode: TypeMappingMode): PsiType
 }
 
 public interface KtPsiTypeProviderMixIn : KtAnalysisSessionMixIn {
-    public fun KtExpression.getPsiType(mode: TypeMappingMode = TypeMappingMode.DEFAULT): PsiType =
-        analysisSession.psiTypeProvider.getPsiTypeForKtExpression(this, mode)
+    public fun commonSuperType(
+        types: Collection<KtType>,
+        context: KtExpression,
+        mode: TypeMappingMode = TypeMappingMode.DEFAULT
+    ): PsiType? =
```

Comment:
What do you think about generalizing `commonSuperType` for`KtType`s too?

Reply:
Aha, not sure why I didn't do that in the beginning.